File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-0.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> memo(n+1);
        if(cost.size()<=1) return 0;
        if(cost.size()==2) return min(cost[0],cost[1]);
        memo[0]=0;
        memo[1]=0;
        

        for(int i=2;i<=n;i++){
            memo[i]= min(memo[i-1]+ cost[i-1], memo[i-2]+ cost[i-2]);
        }
        return memo[n];
    }
};

File: Data Structures & Algorithms/edit-distance/submission-0.java
class Solution {
    static int[][] dp = new int[100][100];
    public int minDistance(String word1, String word2) {
        for (int i = 0; i < dp.length; i++) {
            Arrays.fill(dp[i], -1);
        }
        
        return(worker(word1,word2,word1.length()-1,word2.length()-1));
    }
    public static int worker(String word1, String word2, int i ,int j){

        if(i==-1){return j+1;}
        if(j==-1){return i+1;}


        if(dp[i][j]!=-1){
            return dp[i][j];
        }

        int match = worker(word1, word2, i-1, j-1);
        if(word1.charAt(i)!=word2.charAt(j)){match++;}

        int del = worker(word1, word2, i, j-1) + 1;

        int add = worker(word1, word2, i-1, j) + 1;

        dp[i][j] = Math.min(Math.min(del ,add),match);
        
        return dp[i][j];
    }
}
